### DISCUSSION TITLE
Allow manual control of the current locale for custom servers

### DISCUSSION THREAD

**Discussion by pdusen** [vercel/next.js#25253] [Ideas]
### Describe the feature you'd like to request

Our application authenticates users by sending them to a separately-hosted authentication portal, which creates an authentication cookie and then sends the user back to our application. Once they come back to us, our express middleware uses the authentication cookie to look up any necessary user data, including what locale we should be using for this particular user.

As a rule, we do not put the current locale in the URL. Instead, we generally load the appropriate locale data using our express middleware, and then the application framework uses that data to display the appropriate content. This has been a point of conflict in the development of our new application, because Next.js appears to tightly couple the user's current locale to the URL, and we haven't found a good workaround to allow us to override it using the request data we have available.

I propose that Next.js be refactored to allow developers to configure custom logic for deducing the appropriate locale to use for any given request (for either server-side or static data).

Note that the possible locales themselves would still be defined as they currently are on the `i18n` property in `next.config.js`.

### Describe the solution you'd like

One possible approach would be to allow developers to define a locale override function in `next.config.js`, which would then be run by Next.js when digesting a request and determining what content to serve. The function could look something like this:

```
module.exports = {
  // ...
  i18n: {
    locales: ['en-US', 'de-DE'],
    defaultLocale: 'en-US',
    localeDetection: false,
    overrideLocale: (req) => {
      return req.user.locale;
    },
  },
  // ...
}
```

Alternatively, the function could be a new possible value for `localeDetection` itself, if this logic implies that the current redirect logic for `localeDetection` as it currently exists isn't supposed to be used. In that case, the configuration could be written like this:

```
module.exports = {
  // ...
  i18n: {
    locales: ['en-US', 'de-DE'],
    defaultLocale: 'en-US',
    localeDetection: (req) => {
      return req.user.locale;
    },
  },
  // ...
}
```

### Describe alternatives you've considered

We've tried to work around the absence of a locale control feature in a few different ways, without any real success.

We tried making use of [the NEXT_LOCALE cookie](https://nextjs.org/docs/advanced-features/i18n-routing#leveraging-the-next_locale-cookie) by creating a middleware to manually set the cookie on every request based on the loaded locale data before running the Next.js request handler. However, it quickly became evident that the cookie only works when `localeDetection` is enabled, which doesn't work for us since we don't want the locale to be part of the page URL.

[This issue comment ](https://github.com/vercel/next.js/issues/20761#issuecomment-821842116) seems to imply that it's possible to force a particular locale's content to be used by creating a `__nextLocale` query parameter on the request. We attempted to do this using a middleware similarly to the previous attempt; however, in debugging, it became clear that this query parameter is always overwritten by a value from Next.js' own locale detection logic, so this didn't work either.

**Comment by vitalybaev** [comment]
Maybe it would be simpler just to rewrite `Accept-Language` header in your server middleware?

**Reply by pdusen** [reply]
> Maybe it would be simpler just to rewrite `Accept-Language` header in your server middleware?

I can attempt this, but judging by the documentation, I expect Next.js' behavior in that case to be the same as for the `NEXT_LOCALE` cookie; to do nothing if `localeDetection` is disabled, and to try and redirect the user to a page with the locale in the URL if it's enabled, which is precisely what we want to avoid.

Though even if that works, it would still be a workaround, and certainly not "simpler" than having direct control of locale detection in the Next.js configuration itself.

**Reply by pdusen** [reply]
Confirmed my above assumption; the behavior when overriding the `Accept-Language` header is the same as for overriding the `NEXT_LOCALE` cookie.

**Comment by sudheeshcm** [comment]
@pdusen Hey, 
I'm also trying for the same implementation. Would like to have locale implementation without modifying the URLs.
Please do post your finding if you were able to solve this.

My next js App is able to translate the contents from the client side when passing the selected language as a prop to the translate method. But it's not always feasible to pass in the locale as a prop for all translate method usage.
Tried setting the `NEXT_LOCALE` cookie too. That too didn't have any effect on the any of the rendered pages.

**Reply by pdusen** [reply]
The only solution I've found for the moment is to avoid Next.js locale support entirely and handle it manually. It's far from ideal, but it's all we have at the moment.

**Reply by sudheeshcm** [reply]
@pdusen 
I'm currently trying this workaround.

**Code to figure out the locale from cookies (NEXT_LOCALE)**
```
export async function getServerSideProps({ locale, req }: GetServerSidePropsContext) {
const cookieLocale = req.cookies && req.cookies['NEXT_LOCALE']
   return {
       props: {
           ...(await serverSideTranslations(cookieLocale || locale || 'en')),
       },
   }
}
```

Above approach reads the cookie and sets the translations as per the locale cookie.
If the strategy is to avoid using URL for locale determination, the Router will not be able to figure out the selected locale.

Workaround for the above issue:
```
function Page(props) {
    const { _nextI18Next: i18n } = props
    const initialLocale = i18n.initialLocale
```

I'm planning for this sort of usage. The next.js locale will never be set correctly. But this approach will work fine.

**Comment by zomars** [comment]
I was wondering the same things, we already have the locale saved in a DB for the user. So once he's logged in we want to display the dashboard in their language without changing URLs. I would like to use `useRouter` locale but it will always return `en` (default).

**Comment by bertho-zero** [comment]
It's really annoying that you can't just define custom logic for language detection
